Plots (#1186)
iterative/dvc.org: content/docs/command-reference/plot/index.md

# plot

> Contains commands to visualize users data stored in structured files like JSON,
> CSV, TSV: [show](/doc/command-reference/plot/show).

Reviewer: ```suggestion
Contains commands to visualize
[continuous metrics](/doc/command-reference/plot#continuous-metrics) in
structured files like JSON, CSV, TSV:
[show](/doc/command-reference/plot/show),
```

[diff](/doc/command-reference/plot/diff).

## Synopsis

```usage

> usage: dvc plot [-h] [-q | -v] {show,diff} ...
> positional arguments:
>   {show,diff}    Use `dvc plot CMD --help` to display command-specific help.
>     show         Plot data from a file

Reviewer: ```suggestion
usage: dvc plot [-h] [-q | -v] {show,diff} ...

positional arguments:
  COMMAND
    show         Plot data from a file
```

    diff         Plot changes between commits in the DVC repository,
                 or between the last commit and the workspace.
```

## Description

> DVC provides a set of commands to visualize
> [continuous metrics](/doc/command-reference/plot#continous-metrics) of machine
> learning experiments in <abbr>DVC projects</abbr>. These metrics are usually
> plotted as AUC curves, loss functions, confusion matrixes and other types of
> plots.

Reviewer: ```suggestion
DVC provides a set of commands to visualize continuous metrics of machine
learning experiments in. Usual examples of plots are AUC curves, loss
functions, and confusion matrices.
```

The continuous metrics should be saved in files which are usually created by
users or generated by user's modeling or data processing code. The plot commands
can work with these files commited to a repository history, data files
controlled by DVC or files from workspace.

> ### Continous metrics

Reviewer: ```suggestion
### Continuous metrics
```

DVC has two concepts for metrics for representing result of machine learning
training or data processing:

1. `dvc metrics` to represent scalar numbers such as AUC, true positive rate and
   others.
2. `dvc plot` to visualize continuous metrics such as AUC curve, loss function,

>    confusion matrixes and others.

Reviewer: Question: maybe we should use "scalar metrics" everywhere i.e. in the dvc metrics commands and other places where they're currently mentioned? Always try to use either scalar or continuous except in placs where metrics are mentioned in an abstract sense.

Author: > maybe we should use "scalar metrics" everywhere

Yes. It should be a very clear separation.

Reviewer: OK added a checkbox to #1175

# plot

Contains commands to visualize
[continuous metrics](/doc/command-reference/plot#continuous-metrics) in
structured files like JSON, CSV, TSV: [show](/doc/command-reference/plot/show),
[diff](/doc/command-reference/plot/diff).

## Synopsis

```usage
usage: dvc plot [-h] [-q | -v] {show,diff} ...

positional arguments:
  COMMAND

>     show         Plot data from a file
>     diff         Plot changes between commits in the DVC repository,
>                  or between the last commit and the workspace.

Reviewer: ```suggestion
    show         Generate a plot image file from a continuous metrics file
    diff         Plot continuous metrics differences between commits in the DVC repository,
                 or between the last commit and the workspace.
```
> May need rewrapping
Cc @pared also for the core repo PR

Reviewer: Moved to iterative/dvc/pull/3577

Reviewer: The definition for show is still pending in the core repo (https://github.com/iterative/dvc/pull/3577#discussion_r416844217)
But the one for diff has been applied ✅

Reviewer: Oh I see this one has been applied. Thanks, resolving.

```

## Description

> DVC provides a set of commands to visualize continuous metrics of machine

Reviewer: ```suggestion
DVC provides a set of commands to visualize _continuous metrics_ of machine
```

learning experiments in. Usual examples of plots are AUC curves, loss functions,

> and confusion matrices.

Author: All points are great! I rearranged this part. Please take a look.

It is still important to keep the scalars vs continuous difference somewhere. I keep it as a next section but rearranged from the plot point of view.

> and confusion matrices.

Reviewer: Thanks! I think it's easier to read now 🙂 

> ![image](https://user-images.githubusercontent.com/1477535/80649495-c61f5d00-8a37-11ea-8282-c2235633b016.png)

I agree! Unfortunately I pasted the wrong text in my previous message and so I didn't express myself correctly haha. This is what I meant:

![image](https://user-images.githubusercontent.com/1477535/80649484-bdc72200-8a37-11ea-9001-e45b99e148ba.png)

So to summarize, I think the scalar metrics should be explained in the description of `dvc metrics` cmd ref index, and just mention the word here with a link to there, without explaining. But OK I guess that can be a separate issue (to match dvc metrics with this ref)

> and confusion matrices.

Reviewer: Oops I did it again (wrong paste). Please read the comment above in Github ☝️ because I edited it...

> and confusion matrices.

Reviewer: p.s. unresolving for myself to remember about "that can be a separate issue (to match dvc metrics with this ref)"

`dvc metrics diff` command can represent difference between the metrics in
different experiments as a float numbers. Like `AUC` metrics is `0.801807` and
was increase by `+0.037826` from the previous value:

```dvc
$ dvc metrics diff
    Path       Metric    Value      Change
summary.json   AUC      0.801807   0.037826
```

In contrast to scalar metrics, continous metrics represents a plot and should be
stored as an array in JSON file or as a column in CSV or TSV files. The command
`dvc plot diff` generates a plot with two versions of the metrics:

```dvc
$ dvc plot diff -d logs.csv
file:///Users/dmitry/src/plot/logs.html
```

![](/img/plot_auc.svg)

### File formats

Supported file formats for continuous metrics are: JSON, CSV, TSV. DVC expects
to see an array (or multiple arrays) of objects (usually _float numbers_) in the
file.

In tabular file formats such as CSV and TSV the array is a column. Plot command
can generate visuals for a specified column or a set of columns. Like `AUC`
column:

```
epoch, AUC, loss
34, 0.91935, 0.0317345
35, 0.91913, 0.0317829
36, 0.92256, 0.0304632
37, 0.92302, 0.0299015
```

> In hierarchical file formats such as JSON an array of JSON-objects is expected.
> Plot command can generate visuals for a specified field name or a set of fields
> from the array's object. Like `val_loss` field in the `train` array in this
> example:

Reviewer: ```suggestion
In JSON files (hierarchical format) a named array of matching JSON objects is
expected. `dvc plot show` can generate visuals for a one, several, or all field
names from the array's objects. For example `val_loss` in the `train` array below:
```

# plot

Contains commands to visualize
[continuous metrics](/doc/command-reference/plot#continuous-metrics) in
structured files (JSON, CSV, or TSV): [show](/doc/command-reference/plot/show),
[diff](/doc/command-reference/plot/diff).

## Synopsis

```usage
usage: dvc plot [-h] [-q | -v] {show,diff} ...

positional arguments:
  COMMAND
    show         Generate a plot image file from a continuous metrics file.
    diff         Plot continuous metrics differences between commits in the DVC
                 repository, or between the last commit and the workspace.
```

## Description

> DVC provides a set of commands to visualize _continuous metrics_ of machine
> learning experiments. Usual examples of plots are AUC curves, loss functions,
> and confusion matrices.

Reviewer: ```suggestion
Continuous metrics represent plots, and should be stored as data series in one
of the supported [file formats](#file-formats). These files are usually created
by users or generated by user modeling or data processing code.
```

> Continous metrics represents a plot and should be stored as data series in one
> of the supported [file formats](#file-formats). These files are usually created
> by users or generated by user's modeling or data processing code.

Reviewer: ```suggestion
Continuous metrics represent plots, and should be stored as data series in one
of the supported [file formats](#file-formats). These files are usually created
by users or generated by user modeling or data processing code.
```

Reviewer: But how are plots defined? Is there a new option in `dvc run` for these, like -m for metrics? I don't see info about this in any plot ref yet, nor an update on run itself. Does that mean these metrrics are not tracked by DVC like scalar metrics? Probably worth mentioning either way.

Author: @jorgeorpinel there is no plot metrics in `dvc run` yet. we should introduce one. `-p` is taken by params. What do you think about `-l`\`-L` (pLot)?

Reviewer: I like `-l` (`--plots`) since it comes after `-m` (`--metrics`)

Will there also be a upper case `-L` (`--plot-no-cache`)?

All of this should probably be its own issue in the core repo 😬 I can create it if you'd like, please lmk.

[diff](/doc/command-reference/plot/diff).

## Synopsis

```usage
usage: dvc plot [-h] [-q | -v] {show,diff} ...

positional arguments:
  COMMAND
    show         Generate a plot image file from a continuous metrics file.
    diff         Plot continuous metrics differences between commits in the DVC
                 repository, or between the last commit and the workspace.
```

## Description

DVC provides a set of commands to visualize _continuous metrics_ of machine
learning experiments. Usual examples of plots are AUC curves, loss functions,
and confusion matrices.

Continuous metrics represent plots, and should be stored as data series in one
of the supported [file formats](#file-formats). These files are usually created
by users or generated by user modeling or data processing code.

The plot commands can work with these continuous metrics files that are commited
to a repository history, data files controlled by DVC or files from workspace.
For examlpe, the command `dvc plot diff` generates a plot with two versions of
the metrics:

```dvc
$ dvc plot diff -d logs.csv
file:///Users/dmitry/src/plot/logs.html
```

![](/img/plot_auc.svg)

> ### Difference between continuous and scolar metrics

Reviewer 2: scalar

Reviewer: Adding suggestion for quick commit from GH:
```suggestion
### Difference between continuous and scalar metrics
```
